src/rebar_app_discover.erl

```diff
@@ -34,13 +34,15 @@ format_error({missing_module, Module}) ->
     io_lib:format("Module defined in app file missing: ~p~n", [Module]).
 
 merge_deps(AppInfo, State) ->
-    Profiles = rebar_state:current_profiles(State),
+    Default = rebar_state:default(State),
+    CurrentProfiles = rebar_state:current_profiles(State),
     Name = rebar_app_info:name(AppInfo),
     C = rebar_config:consult(rebar_app_info:dir(AppInfo)),
 
+    %% We reset the opts here to default so no profiles are applied multiple times
     AppState = rebar_state:apply_overrides(
                  rebar_state:apply_profiles(
-                   rebar_state:new(State, C, rebar_app_info:dir(AppInfo)), Profiles), Name),
+                   rebar_state:new(rebar_state:opts(State, Default), C, rebar_app_info:dir(AppInfo)), CurrentProfiles), Name),
```

Comment:
What I'm mostly interested in is figuring out how we make sure this doesn't happen again, and that we clearly define what is okay to do to profiles and opts, and what is unsafe to do to profiles and opts.

It's still early enough to make that decision, but we kind of have to make it. I'm having a hard time understanding the current semantics as it is.